### PR TITLE
PIM-8710: New converter to flatten invalid import values

### PR DESCRIPTION
**Pull request by jmleroux**
This PR fixes an error raised when trying to write the invalid data file for an import with warnings.

We tried to import a file with a valid date field and another invalid attribute field.

- the row is read from the XLSX file and the date field is automtically casted into a DateTime object
- the row is invalidated because of the invalid field value
- the row is written in the invalid data file.

BUT, because the date field is still a DateTime, the JSON File buffer [will encode it into an array](https://github.com/akeneo/pim-community-dev/blob/3.0/src/Akeneo/Tool/Component/Buffer/JSONFileBuffer.php#L58) which is not writable.

With this fix, the row is buffered into a into a pure flat key => string value array.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        return $convertedItem;
+    }
+
+    /**
+     * @throws DataArrayConversionException
+     */
+    private function convertToString($property, $data): string
+    {
+        if ($data instanceof \DateTime) {
+            $convertedData = $data->format('Y-m-d H:i:s');
```
Comment (skeleton) [inline src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php]: dunno about that because if a user tries to reimport invalid data, this value won't be as before. maybe it's not a problem, you should check with po

File: src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            $convertedItem[$property] = $this->convertToString($property, $data);
+        }
+
+        return $convertedItem;
+    }
+
+    /**
+     * @throws DataArrayConversionException
+     */
+    private function convertToString($property, $data): string
```
Comment (skeleton) [inline src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php]: string $property

File: src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            $convertedData = implode(',', $data);
+        } elseif (is_object($data) && !method_exists($data, '__toString')) {
+            throw new DataArrayConversionException(
+                    sprintf('The property "%s" could not be converted into string.', $property)
+                );
+        } else {
+            $convertedData = (string) $data;
+        }
+
+        if (!is_string($convertedData)) {
```
Comment (skeleton) [inline src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php]: `if (null === $convertedData)` + initialize variable at the beginning of the method? would be clearer

**Comment by jmleroux** [comment]
> about the naming maybe something like `ConvertInvalidDataToString` ?

No because it is just a converter. Currently i use it only for invalid data, but it's only purpose is to flatten an array into a simpler key-value array.

File: src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php (hunk 1 above)
Comment (jmleroux) [reply src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php]: Good catch, will check.

File: src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php (hunk 3 above)
Comment (jmleroux) [reply src/Akeneo/Tool/Component/Connector/ArrayConverter/StringArrayConverter.php]: I initialized it because it's obviously clearer. But no need to test for `null` : `is_string(null) = false` and its sufficient.

**Comment by skeleton** [comment]
> > about the naming maybe something like `ConvertInvalidDataToString` ?
> 
> No because it is just a converter. Currently i use it only for invalid data, but it's only purpose is to flatten an array into a simpler key-value array.

what I don't like with StringArrayConverter, it's too technical and to precise. If later, you have to add a new conversion in your class, the naming will be wrong. The functional purpose is to handle the invalid data, not to convert data in a classical import way, we already have converter for that.
But I already approved your PR, do what you want ;)

**Comment by jmleroux** [comment]
> The functional purpose is to handle the invalid data, not to convert data in a classical import way, we already have converter for that.

I get your point. This converter is shitty and should not be used elsewhere than to convert invalid data. I will rename. Thx.
